File: player-core/src/main/java/com/libreshockwave/player/debug/Breakpoint.java
package com.libreshockwave.player.debug;

/**
 * Represents a breakpoint at a specific script location.
 * This is an immutable record.
 */
public record Breakpoint(
    int scriptId,
    String handlerName,
    int offset,
    boolean enabled
) {
    /**
     * Creates a simple enabled breakpoint.
     */
    public static Breakpoint simple(int scriptId, String handlerName, int offset) {
        return new Breakpoint(scriptId, handlerName, offset, true);
    }

    /**
     * Return a copy with enabled state changed.
     */
    public Breakpoint withEnabled(boolean enabled) {
        return new Breakpoint(scriptId, handlerName, offset, enabled);
    }

    /**
     * Unique key for this breakpoint (scriptId:handlerName:offset).
     */
    public String key() {
        return scriptId + ":" + handlerName + ":" + offset;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Breakpoint[").append(scriptId).append(":").append(handlerName).append(":").append(offset);
        if (!enabled) sb.append(", disabled");
        sb.append("]");
        return sb.toString();
    }
}

File: cpp/src/player/debug/Breakpoint.cpp
#include "libreshockwave/player/debug/Breakpoint.hpp"

#include <sstream>
#include <utility>

namespace libreshockwave::player::debug {

Breakpoint Breakpoint::simple(int scriptId, std::string handlerName, int offset) {
    return Breakpoint{scriptId, std::move(handlerName), offset, true};
}

Breakpoint Breakpoint::withEnabled(bool newEnabled) const {
    return Breakpoint{scriptId, handlerName, offset, newEnabled};
}

std::string Breakpoint::key() const {
    return std::to_string(scriptId) + ":" + handlerName + ":" + std::to_string(offset);
}

std::string Breakpoint::toString() const {
    std::ostringstream out;
    out << "Breakpoint[" << scriptId << ":" << handlerName << ":" << offset;
    if (!enabled) {
        out << ", disabled";
    }
    out << "]";
    return out.str();
}

BreakpointKey BreakpointKey::of(int scriptId, std::string handlerName, int offset) {
    return BreakpointKey{scriptId, std::move(handlerName), offset};
}

BreakpointKey BreakpointKey::of(const Breakpoint& breakpoint) {
    return BreakpointKey{breakpoint.scriptId, breakpoint.handlerName, breakpoint.offset};
}

std::string BreakpointKey::toString() const {
    return std::to_string(scriptId) + ":" + handlerName + ":" + std::to_string(offset);
}

} // namespace libreshockwave::player::debug
